internal/venice-common/src/main/java/com/linkedin/venice/memory/ClassSizeEstimator.java

```diff
@@ -0,0 +1,245 @@
+    JAVA_MAJOR_VERSION = Utils.getJavaMajorVersion();
+  }
+
+  private ClassSizeEstimator() {
+    // Static utility
+  }
+
+  /**
+   * This function provides the "shallow size" of each instance of the given class, meaning that all pointers are
+   * considered to be null.
+   *
+   * Intended usage: This function is a helper to make it easier to implement {@link Measurable#getHeapSize()}. It
+   * should not be called on the hot path. Rather, it should be called at most once per JVM runtime per class of
+   * interest and the result should be stored in a static field. If an instance of a measured class contains only
+   * primitive fields or all its non-primitive fields are null, then the result of this function is effectively the heap
+   * size of that instance. If these conditions are not true, then a function should be implemented which uses this
+   * class overhead as a base and then adds the size of any referenced Objects. For example, see
+   * {@link InstanceSizeEstimator#getObjectSize(Object)}.
+   *
+   * @param c The {@link Class} for which to predict the "shallow overhead", as defined above.
+   * @return The base overhead of the class, which can be any positive number (including zero).
+   */
+  public static int getClassOverhead(@Nonnull final Class<?> c) {
+    Integer knownSize = KNOWN_SHALLOW_SIZES.get(c);
+    if (knownSize != null) {
+      return knownSize;
+    }
+    return computeClassOverhead(c);
+  }
+
+  private static int computeClassOverhead(@Nonnull final Class<?> c) {
+    if (c == null) {
+      throw new NullPointerException("The class param must not be null.");
+    }
+
+    if (c.isEnum()) {
+      /**
+       * Each enum value is stored statically, and so the per-instance overhead of enum fields is only the pointer size,
+       * already taken into account in the field loop below.
+       */
+      return 0;
+    }
+
+    if (c.isPrimitive()) {
+      return getPrimitiveSize(c);
+    }
+
+    int size = c.isArray() ? ARRAY_HEADER_SIZE : OBJECT_HEADER_SIZE;
+
+    /**
+     * We need to measure the overhead of fields for the passed in class as well as all parents. The order in which we
+     * traverse these classes matters in older Java versions (prior to 15) and not in newer ones, but for the sake of
+     * minimizing code size, we will always iterate in the order relevant to older Java versions.
+     */
+    List<Class> classHierarchyFromSubclassToParent = new ArrayList<>();
+    classHierarchyFromSubclassToParent.add(c);
+    Class parentClass = c.getSuperclass();
+    while (parentClass != null) {
+      classHierarchyFromSubclassToParent.add(parentClass);
+      parentClass = parentClass.getSuperclass();
+    }
+
+    // Iterate from the end to the beginning, so we go from parent to sub
+    for (int i = classHierarchyFromSubclassToParent.size() - 1; i >= 0; i--) {
+      int classFieldsOverhead = overheadOfFields(classHierarchyFromSubclassToParent.get(i));
```

Comment:
Do we want to use the cache here? Some parent class might be calculated before.

Reply:
I think that would be incorrect, because the cache contains the final size including object header and alignment. Here, we are looking for the "intermediate result" (the fields only), so we can add all these intermediate results together and account for alignment afterwards. We could have a separate cache for these intermediate results but I don't think it's worth it given that the intent is to only call this once per class per runtime, never on the hot path...

Reply:
Got it, makes sense.
Can you also convert these explanation into Javadoc?